File: CPlusPlus/5. 最长回文子串.cpp
// 做法 1： 中心开花， 看最长的回文
// 做法 2： 动态规划， dp[i][j] 表示 s[i..j] 的回文长度， 0 则不回文， j-i+1 则为回文长度

#include <string>
#include <vector>

using namespace std;
class Solution {
public:
    string longestPalindrome(string s)
    {
        int res_len { 1 };
        string res { s.substr(0, 1) };
        int n = s.size();
        vector<vector<int>> dp(n, vector<int>(n, 0));
        for (int i = 0; i < n; ++i)
            dp[i][i] = 1;
        for (int len = 2; len <= n; ++len) {
            for (int i = 0; i + len <= n; ++i) {
                int j = i + len - 1;
                if (s[i] == s[j]) {
                    if (len == 2)
                        dp[i][j] = 2;
                    else if (dp[i + 1][j - 1] > 0)
                        dp[i][j] = dp[i + 1][j - 1] + 2;
                    if (dp[i][j] > res_len) {
                        res_len = dp[i][j];
                        res = s.substr(i, len);
                    }
                }
            }
        }
        return res;
    }
};

File: Rust/5. 最长回文子串.rs
// dp[i][j] 表示 i..=j 的回文长度
impl Solution {
    pub fn longest_palindrome(s: String) -> String {
        let len = s.len();
        let sb = s.as_bytes().to_vec();
        let mut dp = vec![vec![0; len]; len];
        for i in 0..len {
            dp[i][i] = 1;
        }
        let mut curr_len = 1usize;
        let mut res = String::new();
        res.push(sb[0] as char);
        for j in 2..=len {
            for i in 0..=len - j {
                if sb[i] == sb[i + j - 1] {
                    if j == 2 {
                        dp[i][i + j - 1] = 2;
                    } else if dp[i + 1][i + j - 2] != 0 {
                        dp[i][i + j - 1] = dp[i + 1][i + j - 2] + 2;
                    }
                }
                if dp[i][i + j - 1] > curr_len {
                    curr_len = dp[i][i + j - 1];
                    res = (&s[i..(i + j)]).to_string();
                }
            }
        }
        res
    }
}
